File: Task/Conjugate-transpose/Python/conjugate-transpose.py
def conjugate_transpose(m):
    return tuple(tuple(n.conjugate() for n in row) for row in zip(*m))

def mmul( ma, mb):
    return tuple(tuple(sum( ea*eb for ea,eb in zip(a,b)) for b in zip(*mb)) for a in ma)

def mi(size):
    'Complex Identity matrix'
    sz = range(size)
    m = [[0 + 0j for i in sz] for j in sz]
    for i in range(size):
        m[i][i] = 1 + 0j
    return tuple(tuple(row) for row in m)

def __allsame(vector):
    first, rest = vector[0], vector[1:]
    return all(i == first for i in rest)

def __allnearsame(vector, eps=1e-14):
    first, rest = vector[0], vector[1:]
    return all(abs(first.real - i.real) < eps and abs(first.imag - i.imag) < eps
               for i in rest)

def isequal(matrices, eps=1e-14):
    'Check any number of matrices for equality within eps'
    x = [len(m) for m in matrices]
    if not __allsame(x): return False
    y = [len(m[0]) for m in matrices]
    if not __allsame(y): return False
    for s in range(x[0]):
        for t in range(y[0]):
            if not __allnearsame([m[s][t] for m in matrices], eps): return False
    return True


def ishermitian(m, ct):
    return isequal([m, ct])

def isnormal(m, ct):
    return isequal([mmul(m, ct), mmul(ct, m)])

def isunitary(m, ct):
    mct, ctm = mmul(m, ct), mmul(ct, m)
    mctx, mcty, cmx, ctmy = len(mct), len(mct[0]), len(ctm), len(ctm[0])
    ident = mi(mctx)
    return isequal([mct, ctm, ident])

def printm(comment, m):
    print(comment)
    fields = [['%g%+gj' % (f.real, f.imag) for f in row] for row in m]
    width = max(max(len(f) for f in row) for row in fields)
    lines = (', '.join('%*s' % (width, f) for f in row) for row in fields)
    print('\n'.join(lines))

if __name__ == '__main__':
    for matrix in [
            ((( 3.000+0.000j), (+2.000+1.000j)),
            (( 2.000-1.000j), (+1.000+0.000j))),

            ((( 1.000+0.000j), (+1.000+0.000j), (+0.000+0.000j)),
            (( 0.000+0.000j), (+1.000+0.000j), (+1.000+0.000j)),
            (( 1.000+0.000j), (+0.000+0.000j), (+1.000+0.000j))),

            ((( 2**0.5/2+0.000j), (+2**0.5/2+0.000j), (+0.000+0.000j)),
            (( 0.000+2**0.5/2j), (+0.000-2**0.5/2j), (+0.000+0.000j)),
            (( 0.000+0.000j), (+0.000+0.000j), (+0.000+1.000j)))]:
        printm('\nMatrix:', matrix)
        ct = conjugate_transpose(matrix)
        printm('Its conjugate transpose:', ct)
        print('Hermitian? %s.' % ishermitian(matrix, ct))
        print('Normal?    %s.' % isnormal(matrix, ct))
        print('Unitary?   %s.' % isunitary(matrix, ct))

File: Task/Conjugate-transpose/R/conjugate-transpose.r
conjugate_transpose<- function(M){
  t(Conj(M))}

is_hermitian <- function(M,eps=10**-10){
  all(abs(M-conjugate_transpose(M))<eps)
}

is_normal <-function(M,eps=10**-10){
  conjugate <- conjugate_transpose(M)
 all(abs(M%*%conjugate- conjugate%*%M)<eps)
}

is_unitary  <-function(M,eps=10**-10){
 all(abs(M%*% conjugate_transpose(M)-diag(nrow(M)))<eps)
}

M1 = matrix(c(3+0i,2-1i,
              2+1i,1+0i),nrow=2)
cat("M1")
cat("\n")
print(M1)
cat("\n")
Conj_M1 = conjugate_transpose(M1)
cat("Conjugate transpose of M1")
cat("\n")
print(Conj_M1)
cat(paste0("Hermitian ? ", is_hermitian(M1)))
cat("\n")
cat(paste0("Normal ? ", is_normal(M1)))
cat("\n")
cat(paste0("Unitary ? ", is_unitary(M1)))


M2 = matrix(c(1+0i,0+0i,1+0i,
              1+0i,1+0i,0+0i,
              0+0i,1+0i,1+0i),nrow=3)
cat("\n")
cat("M2")
cat("\n")
print(M2)
cat("\n")
Conj_M2 = conjugate_transpose(M2)
cat("Conjugate transpose of M2")
cat("\n")
print(Conj_M2)
cat(paste0("Hermitian ? ", is_hermitian(M2)))
cat("\n")
cat(paste0("Normal ? ", is_normal(M2)))
cat("\n")
cat(paste0("Unitary ? ", is_unitary(M2)))


M3 = matrix(c(sqrt(2)/2+0i,0-(sqrt(2)/2)*1i,0+0i,
              sqrt(2)/2+0i,0+(sqrt(2)/2)*1i,0+0i,
              0+0i,0+0i,0+1i),ncol=3)
cat("\n")
cat("M3")
cat("\n")
print(M3)
cat("\n")
Conj_M3 = conjugate_transpose(M3)
cat("Conjugate transpose of M3")
cat("\n")
print(Conj_M3)
cat(paste0("Hermitian ? ", is_hermitian(M3)))
cat("\n")
cat(paste0("Normal ? ", is_normal(M3)))
cat("\n")
cat(paste0("Unitary ? ", is_unitary(M3)))
